### ISSUE TITLE
Push new workers to existing Context?

### ISSUE THREAD

**Issue by DrChainsaw** [JuliaParallel/Dagger.jl#145]
From my understanding, the Context basically plays the role of a WorkerPool. 

Question is if it should be allowed to push new workers to a Context while the scheduler is working on the DAG?

It does not seem to work as is right now as the condition for popping and firing new tasks does not look for new workers in the Context. When adding this to the compute_dag its seems like new workers are being used, but perhaps there are more subtle issues with this approach? This obviously requires fixing #143 as well.

If not I can make a PR for adding this to the existing scheduler.

**Comment by jpsamaroo** [comment]
I definitely can see the value in dynamically modifying the available workers. Feel free to file a PR, I'd be happy to review.

**Comment by DrChainsaw** [comment]
I was thinking of starting this tonight, but it would be good to have #144 to build on. Shall I just add this functionality in that PR since the PR is pretty small to begin with?

I'm also wondering if the scheduler is considered to be very latency sensitive? A bare-bones implementation would just check the Context for new workers, but this requires things like setdiff which adds a couple of us/ms to the scheduling loop depending on how many workers there are. 

A more latency optimized implementation would perhaps even use a separate context type (or maybe scheduler options) for this use case to make the code for checking and adding more workers just never happen if it is not needed. 

Or maybe there is no harm in poping and firing tasks for existing workers even if they happen to be "busy"? The current `if isempty(state.running) && !isempty(state.ready)` does not seem to care about process status, but I'm unsure in what cases it  triggers.

**Comment by jpsamaroo** [comment]
If you can implement more efficient Context serialization according to my latest comment on #144, I'll merge it and you can add this functionality in a new PR (which is preferred).

The scheduler is currently pretty inefficient and not yet optimized, but I'd prefer not to spend any more time doing checks that aren't necessary. What would make the most sense, IMO, is adding a lock within Context that will be taken within `procs(ctx)`. Then you can take the lock from outside the scheduler to modify the processor list, and we can ensure that we don't cache this list within the scheduler.

**Comment by DrChainsaw** [comment]
But won't you need to check for which procs are new and which have already been popped and fired? A new/modified context could hold a separate list of new procs and the scheduler could move them when they are pop-fired. 

Btw, what are the bad things that can happen without the lock? In e.g. java you'd get a ConcurrentModificationException you or someone else pushes to a collection while iterating, but Julia arrays seem to be pretty robust to this.

**Comment by jpsamaroo** [comment]
If you removed a processor that work is already running on, then that work will still need to complete, but new scheduling iterations won't use it for scheduling. This means that there will be some wind-down time for active workers to be removed from the pool, but it's the safest option.

If you add a new processor, it'll get taken into consideration at the next scheduling iteration, because we don't cache which processors are available (we should query from Context every time). If that's not currently the case, then feel free to fix it.

This can work without the lock if you're always running on the same thread, but the moment you aren't, it's possible to race and cause all sorts of issues (including segfaults). Dagger hasn't been designed to be thread-safe so far, but I would like further development to move in the direction of thread safety. The context should be the only thing that the user can easily modify (since the scheduler state is instantiated internally, and the options are immutable), so we need to just lock that to be able to safely modify it.

**Comment by DrChainsaw** [comment]
Makes sense about when removing processors. I wasn't thinkig of addressing it, but after looking through the code more carefully it looks like it is also straight forward to accomodate in the way you described.

The way I picture the design in my head right now, the context becomes some kind of callback hub and I can't help thinking that there is a better design with a more explicit callback mechanism.

Just to make sure I got this right:  There is no explicit tracking of which procs are in use. The Context in its current form holds all procs, but except for the `isempty(state.running) && !isempty(state.ready)` condition (which I still haven't wrapped my head around) the procs are only used before the scheduling while loop starts. Procs which become free are returned together with the result through the channel one by one and reused right away.

The "callback hub" Context would then hold both new processes (or ids) to be injected as well as existing processes to be removed (when they are pulled from the channel) besides the list of all procs. Using new procs or removing existing procs would then move them from the inject/remove arrays to the procs array. Clearly the inject/remove lists does not need to belong with the context. Any injected function which can produce them will do. One advantage of this is that one can easily supply noops with zero overhead. Drawback could be a clunkier API.

**Comment by jpsamaroo** [comment]
What sort of other callbacks would you want in the Context? I don't want to greatly expand Context unless we know of a variety of useful actions that we'd want to proxy through it. Adding and removing workers alone only really needs a lock.

`isempty(state.running) && !isempty(state.ready)` says "If nothing is currently running and we have thunks ready to execute", which is how we do a fast-path to add a bunch of work onto workers. I think it's kind of silly, and when we eventually make the scheduler multithreaded, it will probably be removed in favor of a workqueue system or similar. Anyway, procs is still used later in the scheduling pipeline, and should be queried with `procs(ctx)` everytime it's accessed.

**Comment by DrChainsaw** [comment]
> What sort of other callbacks would you want in the Context? I don't want to greatly expand Context unless we know of a variety of useful actions that we'd want to proxy through it. Adding and removing workers alone only really needs a lock.

I fully agree and I was trying to suggest to not even use it for adding/removing workers.

In addition to a lock, there also needs to be information about new workers as well as workers to remove as `procs(ctx)` just gives a list of processes which may or may not be working on thunks. For example, my naive implementation kept an array `busyprocs` of procs which had been used to fire tasks in the scheduling loop and did `newprocs = setdiff(busyprocs, procs(ctx))`. This sure seems to work fine but it is not very elegant as it ended up with double work and double bookkeeping. 

Doing the same for removal would mean removing procs pulled from the channel which are not in `procs(ctx)` which arguably is not as bad as the add case in itself, but perhaps the user would also like to know when the process is available and this will be less elegant if done through the context.

Sorry for dragging you into a long discussion about this. Perhaps I should just implement something and then it becomes clearer what the characteristics will be. Its getting a bit late now so I'll probably start this tomorrow.

Just a simple sketch of the "handle through procs" approach (`popandfire!` is short for the pop_with_affinity! and fire_task! block): 
```julia
function compute_dag(...)
    ...
    ps = procs(ctx)
    busyps = []
    for p in ps
      popandfire!(ctx, p,...)
      push!(busyps, p)
    end

     # Loop while we still have thunks to execute
    while !isempty(state.ready) || !isempty(state.running)

     newps = setdiff(busyps, procs(ctx))
     for p in newps
          popandfire!(ctx, p,...)
          push!(busyps, p)
     end
     
     ...
     proc, thunk_id, res = take!(chan) # get result of completed thunk
     if proc ∉ procs(ctx)
         removeproc!(ctx, proc) # This should also signal to some callback that proc is available
         continue
     end
     ...
end

```

Here is a simple sketch of the callback approach

```julia
function compute_dag(..., newprocs = () -> [], removeproc = p -> false)
     ... 
    # Loop while we still have thunks to execute
    while !isempty(state.ready) || !isempty(state.running)
    ...
    # This can be done in many ways, some are probably better than others...
    for p in newprocs()
        push!(ctx, p)
        popandfire!(ctx, p,...)
    end

    ...
    proc, thunk_id, res = take!(chan) # get result of completed thunk
    if removeproc(proc) # Note that this is also a natural callback to user for proc being available
         removeproc!(ctx, proc)
         continue
    end
    ...
end
```
 The latter is just a bit cleaner imo, except maybe for the extra inputs to compute_dag (which could come as options though).

**Comment by jpsamaroo** [comment]
Why do you need `removeproc!(ctx, proc)` when the proc is already removed from `ctx`? Are you trying to completely remove the worker from Julia's Distributed tracking? If so, I can definitely understand where you're coming from, but it will be more work than you've detailed here, since workers cache data for future thunks. If you want to completely pull everything off a worker, you should start by keeping track of which worker each thunk in `state.running` is assigned to (since I have open PRs that can assign multiple thunks to a single worker). When you remove a proc from `ctx.procs`, you then wait until that worker has no further thunks running on it, and then you need to either wait until all data in `state.cache` is on only other workers, or implement explicit data movement from that to-be-removed worker to an active worker. The latter might require locking down the scheduler until all internal state has been updated, otherwise stale cache handles might remain.

**Comment by DrChainsaw** [comment]
Nope, it is because it does not matter if `proc` is in `procs(ctx)` or not. Once its current task is ready it will be rescheduled. The mechanism for this is when popping it out of the channel (the `take!` line). 

Just to make sure, I did it the TDD way and first just removed `proc` from `ctx.procs` and the testcase did not pass until I also added `&& !shall_remove_proc(ctx, proc,...)` after `finalize_task!`.

**Comment by jpsamaroo** [comment]
Good point, I had forgotten about `immediate_next` (a mechanism which I plan to remove once we're scheduling sub-DAGs).
